### PR TITLE
Add CSVHistory callback to store history metrics in CSV file.

### PR DESCRIPTION
**Pull request by gw0**
Having a `History` callback is great, but the results are available only when the training is finished (which could take quite a while) and is destroyed if the process is killed. The `CSVHistory` callback stores all history metrics in a CSV file at each epoch end, so that you can examine it whenever you want and no matter what happens.

In combination with `ModelCheckpoint` callback and `model.load_weights('foo.hdf5')` it can even be used to resume training if the process is killed.

### CODE REVIEW

**Comment by fchollet** [comment]
I think this is a typical example of something that is useful for some users, and benefits from being shared on Github, but where there is little justification for including it in the master codebase. 

I don't believe this is 1) useful to a sufficiently large fraction of the user base and 2) tightly related to the code functionality of Keras. What do you think?

**Comment by gw0** [comment]
You are right that it is not an essential part of Keras, but I think that everyone who is trying to do more complex research will need to implement something like this. Something to store training progress to a file (possibly on the fly, not just on the end) to analyze it. Possibly together with a plotting callback (I haven't submitted a PR yet) to quickly determine if it is overfitting and how each variation behaves.

**Comment by pasky** [comment]
I agree that everyone serious wants some way to monitor and analyze training progress, but if @fchollet wants to keep the number of callbacks reasonably small, wouldn't it be better to focus all such features to visualization solutions, like just being discussed in #1680? Such a tool could export training progress to CSV or other formats, as well as offering realtime plots, show accuracy per some sort of sample buckets (like sentence lengths), and visualizing parameters and whatnot.  If people had an easy option to do this, how many would prefer to rather *only* save progress to CSV?

So, maybe that's a better place where to cluster features like this; Keras would then benefit from a detailed step-by-step guide to set up such a visualization tool, and having better such tools (+ homepages with screenshots etc.).

Another matter then is restoring the training history from a CSV file, but that does seem like a pretty obscure usecase to me.

**Comment by fchollet** [comment]
> So, maybe that's a better place where to cluster features like this

Yes. Basically, this belongs in a Keras extensions package, e.g. a visualization package.

**Comment by gw0** [comment]
@pasky This was a step towards visualization, because I extended this callback class with ~20 lines of `matplotlib` code to also continuously plot all metrics/parameters, but submitted a PR for only the CSV part to see your feedback.

Ok, if I understand correctly, custom callbacks or other extensions belong in non-official repositories? And the preferred monitoring, analysis and visualization solution is the Bokeh server (or what will be determined by #1680)? It would be great to have something more than a CSV file and simple plots.
